src/main/java/ca/corefacility/bioinformatics/irida/config/web/OpenAPIConfig.java

```diff
@@ -0,0 +1,35 @@
+package ca.corefacility.bioinformatics.irida.config.web;
+
+import io.swagger.v3.oas.annotations.OpenAPIDefinition;
+import io.swagger.v3.oas.annotations.enums.SecuritySchemeType;
+import io.swagger.v3.oas.annotations.info.Contact;
+import io.swagger.v3.oas.annotations.info.Info;
+import io.swagger.v3.oas.annotations.security.*;
+import org.springframework.context.annotation.*;
+import org.springframework.web.servlet.config.annotation.EnableWebMvc;
+import org.springframework.web.servlet.config.annotation.WebMvcConfigurer;
+
+/**
+ * Configuration for IRIDA REST API documentation.
+ */
+@Configuration
+@EnableWebMvc
+@ComponentScan(basePackages = {"org.springdoc"})
+@Import({org.springdoc.core.SpringDocConfiguration.class,
+        org.springdoc.webmvc.core.SpringDocWebMvcConfiguration.class,
+        org.springdoc.core.SwaggerUiConfigProperties.class,
+        org.springdoc.core.SwaggerUiOAuthProperties.class,
+        org.springframework.boot.autoconfigure.jackson.JacksonAutoConfiguration.class})
+@SecurityScheme(
+        name = "oauth2",
+        type = SecuritySchemeType.OAUTH2,
+        flows = @OAuthFlows(authorizationCode = @OAuthFlow(authorizationUrl = "/api/oauth/authorize", tokenUrl = "/api/oauth/token",
+                scopes = {@OAuthScope(name = "read", description = "This is the read scope."), @OAuthScope(name = "write", description = "This is the write scope.")}),
+                password = @OAuthFlow(tokenUrl = "/api/oauth/token")))
+@OpenAPIDefinition(
+        info = @Info(title = "IRIDA REST API", version = "${irida.version}",
+                description = "The IRIDA REST API follows a standard output format, regardless of the resource being accessed. Resources can be accessed as an individual resource or as part of a resource collection.",
+                contact = @Contact(name = "the Bioinformatics Team", email = "helpdesk@cscscience.ca")),
+        security = @SecurityRequirement(name = "oauth2", scopes = {"read", "write"}))
+public class OpenAPIConfig implements WebMvcConfigurer {
```

Comment:
I'm sorry, I'm probably a little confused. Maybe I don't know what the different profiles are and how they're used. I understand if we don't want to run Swagger in IRIDA, because of security concerns. We took it out of IRIDA and run a stand alone version from Github Pages on IRIDA's documentation site. We would then only be turning OpenAPI off in production. I understand if we would want to turn the Ajax API service definition off in production, because it's internal to IRIDA. We would probably want internal documentation of the Ajax API only available to developers in development. But IRIDA's rest API is accessible to users. The IRIDA uploader depends on it and we encourage users to write custom scripts with it. I feel The Open API and Rest API should be treated as a pair. The Open API json file is a service definition file that describes that particular instance of the Rest API. If the Rest API can live in every profile, so should it's service definition. That way when we're trying to use an instance of the Rest API, we can generate any request or response objects given the schema described in the Open API json file. We know that the service definition is up-to-date and belongs with that particular instance of the Rest API. Users can do this with Postman or any programming language of their choice. For example, in Java, you can download a jar that generates classes automatically for all the requests and responses when this file is read. The service definition file is very important for usability. It's purpose goes beyond just the Swagger documentation page. I understand not wanting to run Swagger in production, I don't know if I understand why we would want to turn OpenAPI off in production. Can keep the profiles the same for both the Open API and the Rest API?

Reply:
Please see ec0a6612200dd2cf32cecacc1453f9097145aa82.